engine/access/rpc/backend/backend_stream_transactions.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/access/subscription"
+	"github.com/onflow/flow-go/engine/common/rpc"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+// backendSubscribeTransactions handles transaction subscriptions.
+type backendSubscribeTransactions struct {
+	txLocalDataProvider *TransactionsLocalDataProvider
+	executionResults    storage.ExecutionResults
+	log                 zerolog.Logger
+	broadcaster         *engine.Broadcaster
+	sendTimeout         time.Duration
+	responseLimit       float64
+	sendBufferSize      int
+
+	blockTracker subscription.BlockTracker
+}
+
+// TransactionSubscriptionMetadata holds data representing the status state for each transaction subscription.
+type TransactionSubscriptionMetadata struct {
+	txID               flow.Identifier
+	txReferenceBlockID flow.Identifier
+	messageIndex       counters.StrictMonotonousCounter
+	blockWithTx        *flow.Header
+	txExecuted         bool
+}
+
+// SubscribeTransactionStatuses subscribes to transaction status changes starting from the transaction reference block ID.
+// If invalid tx parameters will be supplied SubscribeTransactionStatuses will return a failed subscription.
+func (b *backendSubscribeTransactions) SubscribeTransactionStatuses(ctx context.Context, tx *flow.TransactionBody) subscription.Subscription {
+	nextHeight, err := b.blockTracker.GetStartHeightFromBlockID(tx.ReferenceBlockID)
+	if err != nil {
+		return subscription.NewFailedSubscription(err, "could not get start height")
+	}
+
+	txInfo := TransactionSubscriptionMetadata{
+		txID:               tx.ID(),
+		txReferenceBlockID: tx.ReferenceBlockID,
+		messageIndex:       counters.NewMonotonousCounter(0),
+		blockWithTx:        nil,
+	}
+
+	sub := subscription.NewHeightBasedSubscription(
+		b.sendBufferSize,
+		nextHeight,
+		b.getTransactionStatusResponse(&txInfo),
+	)
+
+	go subscription.NewStreamer(b.log, b.broadcaster, b.sendTimeout, b.responseLimit, sub).Stream(ctx)
+
+	return sub
+}
+
+// getTransactionStatusResponse returns a callback function that produces transaction status
+// subscription responses based on new blocks.
+func (b *backendSubscribeTransactions) getTransactionStatusResponse(txInfo *TransactionSubscriptionMetadata) func(context.Context, uint64) (interface{}, error) {
+	return func(ctx context.Context, height uint64) (interface{}, error) {
+		highestHeight, err := b.blockTracker.GetHighestHeight(flow.BlockStatusFinalized)
+		if err != nil {
+			return nil, fmt.Errorf("could not get highest height for block %d: %w", height, err)
+		}
+
+		// Fail early if no block finalized notification has been received for the given height.
+		// Note: It's possible that the block is locally finalized before the notification is
+		// received. This ensures a consistent view is available to all streams.
+		if height > highestHeight {
+			return nil, fmt.Errorf("block %d is not available yet: %w", height, storage.ErrNotFound)
```

Comment:
Following on my previous comment, I think we should add a new sentinel to handle this error case instead of overriding `storage.ErrNotFound`.

Let's add `ErrBlockNotReady` to indicate that the subscription should retry the height when it is next run. You can add it to this list of errors here:
https://github.com/onflow/flow-go/blob/00e4924e0d1b2d69223e5db722b8e6f15ce92399/engine/access/subscription/streamer.go#L92-L92